### ISSUE TITLE
torch.ops.spyre.overwrite produces incorrect results

### ISSUE THREAD

**Issue by raghukiran1224** [torch-spyre/torch-spyre#1765]
## Summary

`torch.ops.spyre.overwrite` produces incorrect results when overwriting a single position in a 4D tensor. The written position shows ~2000x more error than untouched positions (max_diff ~4.4 vs 0.002).

This is the same class of bug as #1760 (`torch.cat` with size-1 concat dim), but affecting the `overwrite` custom op used for KV cache fill-mode updates in autoregressive generation.

## Minimal Reproducer

```python
"""spyre.overwrite produces wrong results on the written position."""
import torch
import torch_spyre

torch.manual_seed(42)

cache = torch.randn(1, 8, 64, 128, dtype=torch.float16)
new_val = torch.randn(1, 8, 1, 128, dtype=torch.float16)

# CPU reference
ref = cache.clone()
ref[:, :, 0:1, :] = new_val

# Spyre overwrite
cache_sp = cache.to("spyre")
new_val_sp = new_val.to("spyre")
torch.ops.spyre.overwrite(input=new_val_sp, output=cache_sp, dims=[2], offsets=[0])
out = cache_sp.to("cpu")

diff = (ref - out).abs()
print(f"overwrite @0:  max_diff={diff.max().item():.4f}")
print(f"  written row: max_diff={diff[:,:,0:1,:].max().item():.4f}")
print(f"  other rows:  max_diff={diff[:,:,1:,:].max().item():.4f}")
```

### Output

```
overwrite @0:  max_diff=4.4102
  written row: max_diff=4.4102
  other rows:  max_diff=0.0020
```

## Additional Data

Sequential overwrites accumulate error:

```
overwrite @0:  max_diff=4.4102  mean_diff=0.017838
overwrite @1:  max_diff=4.4766  mean_diff=0.036255
overwrite @2:  max_diff=4.7109  mean_diff=0.053467
...
overwrite @9:  max_diff=5.2969  mean_diff=0.176758
```

The untouched region always shows the expected 0.002 stickify precision — only the overwritten positions are wrong.

## Impact

This affects KV cache fill-mode updates in autoregressive LLM generation on Spyre. The `kv_cache_update` function uses `spyre.overwrite` to write individual K/V slices into the cache during fill steps. The corrupted data causes decode-step token divergence from CPU reference.

## Environment

- torch-spyre: `223312d` (upstream main)
- torch: 2.10.0+cpu
- Shape: `(1, 8, 64, 128)` fp16, overwriting at dim=2

## Related

- #1760 — `torch.cat` size-1 concat dimension (fixed on main)
- #1719 — `torch.cat` numerical mismatch in eager mode

**Comment by yoheiueda** [comment]
Thank you for reporting this issue.

At the moment, support for size‑1 dimensions is limited. In particular, the overwrite handling does not work correctly when the outermost dimension has size 1.

I am currently working on fixing this limitation in:

- #1548

I will follow up on this issue once the implementation in #1548 is complete.

**Comment by raghukiran1224** [comment]
Adding a stronger reproducer from the HF adapter decode path.

The original issue shows `torch.ops.spyre.overwrite` is wrong in a simple eager overwrite. I now see the same failure pattern in a compiled decode-shaped graph that mirrors KV-cache writes in LLM generation:

`RMSNorm -> linear -> view -> transpose -> KV update`

I used this reproducer on the Spyre pod:

```python
# tests/test_kv_cache_update_decode_repro.py
# compares CPU vs Spyre for:
# 1. expansion via legacy torch.cat
# 2. expansion via overwrite into a fresh buffer
# 3. fill via overwrite into an existing cache tensor
```

Key result:

- Expansion is numerically fine.
  - `old_len=64`, legacy `torch.cat`: whole-cache max diff `0.0098`
  - `old_len=64`, overwrite append: whole-cache max diff `0.0098`
  - `old_len=192`, legacy `torch.cat`: whole-cache max diff `0.0098`
  - `old_len=192`, overwrite append: whole-cache max diff `0.0098`
- Fill overwrite into an existing cache tensor is not fine.
  - `old_len=128, token_index=1, cache_position=65`: written-slot max diff `2.8926`, untouched max diff `0.0020`
  - `old_len=128, token_index=2, cache_position=66`: written-slot max diff `2.6152`, untouched max diff `0.0020`
  - `old_len=256, token_index=63, cache_position=255`: written-slot max diff `2.6973`, untouched max diff `0.0020`

So the corruption is localized to the written row of the existing cache, not the untouched region, and not generic cache growth.

This matches the decode symptom in our HF adapter:
- `decode-1` is an expansion step and matches CPU
- `decode-2` is the first fill step and is where divergence starts

Minimal repro used:

```python
import torch
import torch.nn as nn
import torch_spyre  # noqa

DEVICE = "spyre"
torch.manual_seed(42)

BSZ = 1
SEQ = 64
NUM_KV_HEADS = 8
HEAD_DIM = 128
HIDDEN = 1024
EPS = 1e-6


def rmsnorm(x, weight):
    variance = (x * x).mean(-1, keepdim=True)
    eps_t = torch.full((1,), EPS, dtype=x.dtype, device=x.device)
    return weight * (x * torch.rsqrt(variance + eps_t))


class FusedFillOverwrite(nn.Module):
    def __init__(self, token_index, cache_position):
        super().__init__()
        self.token_index = token_index
        self.cache_position = cache_position
        self.ln_w = nn.Parameter(torch.randn(HIDDEN, dtype=torch.float16))
        self.k_proj = nn.Linear(HIDDEN, NUM_KV_HEADS * HEAD_DIM, bias=False)

    def forward(self, x, key_cache):
        h = rmsnorm(x, self.ln_w)
        bsz, sl, _ = x.shape
        k = self.k_proj(h).view(bsz, sl, NUM_KV_HEADS, HEAD_DIM).transpose(1, 2)
        k_slice = k[:, :, self.token_index:self.token_index + 1, :]
        result = torch.ops.spyre.overwrite(
            input=k_slice, output=key_cache, dims=[2], offsets=[self.cache_position],
        )
        return key_cache if result is None else result


def cpu_ref(mod, x, key_cache):
    with torch.no_grad():
        h = rmsnorm(x, mod.ln_w)
        bsz, sl, _ = x.shape
        k = mod.k_proj(h).view(bsz, sl, NUM_KV_HEADS, HEAD_DIM).transpose(1, 2)
    out = key_cache.clone()
    ti = mod.token_index
    pos = mod.cache_position
    out[:, :, pos:pos + 1, :] = k[:, :, ti:ti + 1, :]
    return out


x = torch.randn(BSZ, SEQ, HIDDEN, dtype=torch.float16)
key_cache = torch.randn(BSZ, NUM_KV_HEADS, 128, HEAD_DIM, dtype=torch.float16)
mod = FusedFillOverwrite(token_index=1, cache_position=65).to(torch.float16).eval()
mod.requires_grad_(False)

ref = cpu_ref(mod, x, key_cache)
compiled = torch.compile(mod.to(DEVICE), dynamic=False)
out = compiled(x.to(DEVICE), key_cache.to(DEVICE)).to("cpu")

diff = (ref.float() - out.float()).abs()
print("written row max diff:", diff[:, :, 65:66, :].max().item())
print("untouched max diff:", torch.cat((diff[:, :, :65, :], diff[:, :, 66:, :]), dim=2).max().item())
```

Observed output on the pod:

```text
written row max diff: ~2.89
untouched max diff: ~0.002
```

This seems narrower than a generic overwrite problem:
- overwrite into a fresh destination buffer is fine for expansion-like append
- overwrite into an existing populated cache tensor is wrong for fill-like updates

If helpful I can follow up with the full repro file and exact environment details from the pod.

**Comment by michihirohorie** [comment]
Since overwrite is originally designed to perform the in‑place update, current behavior is within the intended design scope.
Do we need an out‑place update version of overwrite?
